pkg/store/proxy_merge.go

```diff
@@ -227,8 +228,16 @@ type lazyRespSet struct {
 	frameTimeout   time.Duration
 
 	// Internal bookkeeping.
-	dataOrFinishEvent    *sync.Cond
+	dataOrFinishEvent *sync.Cond
```

Comment:
These two `sync.Cond` use the same mutex; `closed` and `noMoreData` also +/- have the same function. Could we simplify these? Or could we move the management of `bufferedResponses` with `bufferSlotEvent`, `fixedBufferSize`, `ring*`, etc., into a separate struct? It is very hard to understand and there are probably some hidden footguns.

Reply:
> These two sync.Cond use the same mutex;

It's totally normal for two condition variables to use the same lock. A condition variable is a one-way message-passing synchronization mechanism. The lock is to protect some data fields. Two condition variables need to read/write those data fields, and therefore hold the lock.

> Could we simplify these?

Do you have any rough ideas for how to simplify them? Or moving the management of the ring buffer out of the struct is enough?

> there are probably some hidden footguns.

Any specific concerns/footguns I can address?

Reply:
I feel that separating the logic of managing the circular buffer into a separate struct would help a lot. I worry that there might be some logic bug we are missing due to the interleaved logic. We previously had some bugs in this area and it's a very important part so I would appreciate if we could separate the logic into a separate struct.

I also thought that maybe https://github.com/thanos-io/thanos/blob/26f6e643655f98b5aa617d9f439df43d00decc29/pkg/store/proxy_test.go#L808 these tests could help catch some bugs but I see that they are broken on main :/